packages/tools/devtools/devtools-core/src/BaseDevtools.ts

```diff
@@ -0,0 +1,450 @@
+import type { ContainerDevtoolsFeatureFlags } from "./Features.js";
+import type { IContainerDevtools } from "./IContainerDevtools.js";
+import type { AudienceChangeLogEntry, ConnectionStateChangeLogEntry } from "./Logs.js";
+import {
+	DataVisualizerGraph,
+	type FluidObjectNode,
+	type RootHandleNode,
+	defaultVisualizers,
+} from "./data-visualization/index.js";
+import {
+	AudienceSummary,
+	ContainerDevtoolsFeatures,
+	ContainerStateChange,
+	ContainerStateHistory,
+	DataVisualization,
+	GetAudienceSummary,
+	GetContainerDevtoolsFeatures,
+	GetContainerState,
+	GetDataVisualization,
+	GetRootDataVisualizations,
+	type IDevtoolsMessage,
+	type ISourcedDevtoolsMessage,
+	type InboundHandlers,
+	type MessageLoggingOptions,
+	RootDataVisualizations,
+	handleIncomingWindowMessage,
+	postMessagesToWindow,
+} from "./messaging/index.js";
+
+/**
+ * Abstract base class for devtools implementations.
+ * Contains shared functionality between ContainerDevtools and DataObjectDevtools.
+ *
+ * @internal
+ */
+export abstract class BaseDevtools implements IContainerDevtools, HasContainerKey {
+	/**
+	 * {@inheritDoc HasContainerKey.containerKey}
+	 */
+	public readonly containerKey: ContainerKey;
+
+	/**
+	 * Data contents of the Container.
+	 *
+	 * @remarks
+	 *
+	 * This map is assumed to be immutable. The devtools will not make any modifications to its contents.
+	 */
+	public readonly containerData?: Record<string, IFluidLoadable>;
+
+	// #region Accumulated log state
+
+	/**
+	 * Accumulated data for {@link IContainerDevtools.getContainerConnectionLog}.
+	 */
+	protected readonly _connectionStateLog: ConnectionStateChangeLogEntry[];
+
+	/**
+	 * Accumulated data for {@link IContainerDevtools.getAudienceHistory}.
+	 */
+	protected readonly _audienceChangeLog: AudienceChangeLogEntry[];
+
+	// #endregion
+
+	/**
+	 * Manages state visualization for containerData, if any was provided.
+	 *
+	 * @remarks Will only be `undefined` if `containerData` was not provided, or if the devtools has been disposed.
+	 */
+	protected dataVisualizer: DataVisualizerGraph | undefined;
+
+	/**
+	 * Whether or not the instance has been disposed yet.
+	 *
+	 * @remarks Not related to Container disposal.
+	 *
+	 * @see {@link IContainerDevtools.dispose}
+	 */
+	protected _disposed: boolean;
+
+	// #region Abstract methods that must be implemented by subclasses
+
+	/**
+	 * Gets the audience associated with this devtools instance.
+	 */
+	public abstract get audience(): IAudience;
+
+	/**
+	 * Gets the set of features supported by this instance.
+	 */
+	protected abstract getSupportedFeatures(): ContainerDevtoolsFeatureFlags;
+
+	/**
+	 * Generates state metadata describing the current state of the associated Container or DataObject.
+	 */
+	protected abstract getContainerState(): ContainerStateMetadata;
+
+	/**
+	 * Gets the message handlers specific to this devtools type.
+	 */
+	protected abstract getSpecificInboundMessageHandlers(): InboundHandlers;
```

Comment:
Rather than getting these via an abstract property, would it make sense to take these in via the constructor? We bind all of the handlers at construction, so we should have all the necessary info at that point.